### PR TITLE
Overlord Initial Submit

### PR DESCRIPTION
**Pull request by jklemen**

### CODE REVIEW

File: overlord/public/js/index.js [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+    for(var k=0; k<disarm_code.length; k++) {
+      var dch = disarm_code.charAt(k);
+      if(isNaN(parseInt(dch))) {
+        $('.flash').html(msg);
+        return;
+      }
+    }
+    if(arm_code.length == 0)
+      arm_code = '1234';
+    if(disarm_code.length == 0)
```
Comment (jklemen) [inline overlord/public/js/index.js]: Hound seems to hate my js files o_O If any reviewers could point me at the style guide it uses for JS I could clean these up.

File: overlord/public/js/jquery.js [hunk 2]
```diff
@@ -0,0 +1,4 @@
...
...he[e];if("string"==typeof b)f[b]=c;else if(n.isEmptyObject(f))n.extend(this.cache[e],b);else for(d in b)f[d]=b[d];return f},get:function(a,b){var c=this.cache[this.key(a)];return void 0===b?c:c[b]},access:function(a,b,c){var d;return void 0===b||b&&"string"==typeof b&&void 0===c?(d=this.get(a,b),void 0!==d?d:this.get(a,n.camelCase(b))):(this.set(a,b,c),void 0!==c?c:b)},remove:function(a,b){var c,d,e,f=this.key(a),g=this.cache[f];if(void 0===b)this.cache[f]={};else{n.isArray(b)?d=b.concat(b.map(n.camelCase)):(e=n.camelCase(b),b in g?d=[b,e]:(d=e,d=d in g?[d]:d.match(E)||[])),c=d.length;while(c--)delete g[d[c]]}},hasData:function(a){return!n.isEmptyObject(this.cache[a[this.expando]]||{})},discard:function(a){a[this.expando]&&delete this.cache[a[this.expando]]}};var L=new K,M=new K,N=/^(?:\{[\w\W]*\}|\[[\w\W]*\])$/,O=/([A-Z])/g;function P(a,b,c){var d;if(void 0===c&&1===a.nodeType)if(d="data-"+b.replace(O,"-$1").toLowerCase(),c=a.getAttribute(d),"string"==typeof c){try{c="true"===c?!0:"false"===c?!1:"null"===c?null:+c+""===c?+c:N.test(c)?n.parseJSON(c):c}catch(e){}M.set(a,b,c)}else c=void 0;return c}n.extend({hasData:function(a){return M.hasData(a)||L.hasData(a)},data:function(a,b,c){
```
Comment (djkotowski) [inline overlord/public/js/jquery.js]: It is beneficial, though not mandatory, to use a CDN to provide jQuery.

File: overlord/public/js/index.js (hunk 1 above)
Comment (djkotowski) [inline overlord/public/js/index.js]: Here's a [CoffeeScript guide](https://github.com/thoughtbot/guides/tree/master/style/coffeescript).  It's not JS, but some of the rules are not CS-specific and can carry over to JS.

File: overlord/classes/bomb.rb [hunk 3]
```diff
@@ -0,0 +1,75 @@
+require './classes/panel.rb'
+
+class Bomb
+  STATES = { 0 => 'armed', 1 => 'disarmed', 2 => 'detonated', 3 => 'hacked' }
+
+  attr_accessor :state, :start_time, :panels
+  attr_accessor :hack_count, :attempts_remain
+
+  def initialize
+    codes = (1...64).to_a.sample(6)
```
Comment (paulghaddad) [inline overlord/classes/bomb.rb]: Consider extracting this into a method.

File: overlord/features/bomb_arming.feature [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+@arm
+Feature: Arming the bomb
+  In order to arm the bomb
+  As an evil overlord
+  I should have to enter the correct code
+
+  @happy
+  Scenario: Arming the bomb with a correct code
+    Given I'm at the bomb page
```
Comment (paulghaddad) [inline overlord/features/bomb_arming.feature]: An overall comment for all your Scenarios is to have fewer steps that are very clear in what each does. For example, you could structure this scenario as follows:

Given I have a disarmed bomb
When I enter the correct arm code
Then the bomb should be armed

For example, that last step definition will contain the logic in the original `Then the bomb should arm` and `And the timer should start` steps.

File: overlord/features/bomb_arming.feature [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+    And I have already submitted my codes
+    And the bomb is disarmed
+    When I enter the correct arm code
+    Then the bomb should arm
+    And the timer should start
+
+  @sad
```
Comment (paulghaddad) [inline overlord/features/bomb_arming.feature]: Good job separating your features into happy and sad paths. You'll do more of this in the Shopping Cart exercise.

File: overlord/features/bomb_code_submit.feature [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+@code-submit
+Feature: Submitting The Arm / Disarm Codes
+  In order to submit codes
+  As the evil overlord
+  I should enter two valid 4-digit numbers
+  And press submit
+
+  @happy
+  Scenario: I enter nothing and arm
+    Given I'm at the bomb page
```
Comment (paulghaddad) [inline overlord/features/bomb_code_submit.feature]: You should be able to extract this step as a `background` step.

File: overlord/features/bomb_code_submit.feature [hunk 7]
```diff
@@ -0,0 +1,80 @@
+#encoding: utf-8
+
+@code-submit
+Feature: Submitting The Arm / Disarm Codes
+  In order to submit codes
+  As the evil overlord
+  I should enter two valid 4-digit numbers
```
Comment (paulghaddad) [inline overlord/features/bomb_code_submit.feature]: This should probably read at a higher level, such as `I should be able to enter arm and disarm codes`. Don't get into the implementation at this point.

File: overlord/features/step_definitions/bomb_panel_codes.rb [hunk 8]
```diff
@@ -0,0 +1,50 @@
+# encoding: utf-8
+require './classes/bomb.rb'
+
+Given(/^an armed bomb$/) do
+  @bomb = Bomb.new
```
Comment (paulghaddad) [inline overlord/features/step_definitions/bomb_panel_codes.rb]: I don't think you should be reaching into your classes for the Cucumber features. These integration tests should only be going through the UI.

File: overlord/overlord.rb [hunk 9]
```diff
@@ -1,14 +1,78 @@
 # run `ruby overlord.rb` to run a webserver for this app
 
+require 'json'
 require 'sinatra'
+require './classes/bomb.rb'
+# require './classes/timer.rb'
```
Comment (paulghaddad) [inline overlord/overlord.rb]: Remove unneeded lines (comments) such as this.

File: overlord/overlord.rb [hunk 10]
```diff
@@ -1,14 +1,78 @@
...
+  # session[:timer] = Timer.new(120)
+  session[:bomb] = Bomb.new
+  haml :index, locals: { bomb: session[:bomb] }
 end
 
-# we can shove stuff into the session cookie YAY!
-def start_time
-  session[:start_time] ||= (Time.now).to_s
+get '/hack' do
+  b = params[:binary].split(//).map(&:to_i)
```
Comment (paulghaddad) [inline overlord/overlord.rb]: I would extract this line and the next into another method. But nice use of the `&.to_i` idiom.

Comment (paulghaddad) [inline overlord/overlord.rb]: Also, I suggest renaming these variables for clarity.

File: overlord/overlord.rb [hunk 11]
```diff
@@ -1,14 +1,78 @@
...
+  p = params[:panel].to_i
+
+  content_type :json
+  { 'success' => session[:bomb].attempt_hack(b, p),
+    'done' => session[:bomb].state == 3 }.to_json
+end
+
+get '/entercode' do
+  success = false
+  if session[:bomb].attempts_remain > 0
```
Comment (paulghaddad) [inline overlord/overlord.rb]: There is a lot going on in this method. The if-else inside an if statement is a smell. Can you simplify or extract part of this method?

File: overlord/overlord.rb [hunk 12]
```diff
@@ -1,14 +1,78 @@
...
+    if session[:bomb].armed?
+      success = session[:bomb].disarm(params[:code])
+    else
+      success = session[:bomb].arm(params[:code])
+    end
+  end
+
+  # session[:timer].toggle if success
```
Comment (paulghaddad) [inline overlord/overlord.rb]: Remove unneeded code.

File: overlord/overlord.rb [hunk 13]
```diff
@@ -1,14 +1,78 @@
...
+
+  content_type :json
+  response = { success: success }
+  response[:attempts_remain] = session[:bomb].attempts_remain
+  response[:state] = session[:bomb].state_to_s
+
+  response.to_json
+end
+
+# get '/timeremain' do
```
Comment (paulghaddad) [inline overlord/overlord.rb]: Remove unneeded code.

File: overlord/overlord.rb [hunk 14]
```diff
@@ -1,14 +1,78 @@
...
+#   response = {}
+#   mins = session[:timer].remain / 60
+#   secs = session[:timer].remain - mins * 60
+#   response[:mins] = mins
+#   response[:secs] = secs
+
+#   response.to_json
+# end
+
+post "/setcodes" do
```
Comment (paulghaddad) [inline overlord/overlord.rb]: There is a lot going on in this method. Consider reworking for simplicity.

**Comment by paulghaddad** [comment]
At this juncture, I recommend a second iteration with several things in mind:

1. Refactor your code for greater simplicity.
2. The Cucumber Integration tests should operate at the UI level. You should not have to pull in the `bomb.rb` class for these tests. These should operate as a user would by clicking links, buttons and UI elements.
3. The `bomb.rb` class should be tested at the method level using RSpec. This should allow you to eliminate some of your integration tests. You'll want to hit all of the edge cases at the unit test level since they are faster.

If you'd like to talk about this in person, I'd be more than happy to help.

**Comment by jklemen** [comment]
Thanks for the feedback!

File: overlord/features/bomb_code_submit.feature (hunk 6 above)
Comment (jklemen) [inline overlord/features/bomb_code_submit.feature]: Hmm interesting. Thanks

File: overlord/features/step_definitions/bomb_panel_codes.rb (hunk 8 above)
Comment (jklemen) [inline overlord/features/step_definitions/bomb_panel_codes.rb]: Some of my tests require verifying functionality that I conveniently already have functions to compute within my class. Is this still not okay? Just wondering.

File: overlord/features/bomb_arming.feature (hunk 4 above)
Comment (jklemen) [inline overlord/features/bomb_arming.feature]: So verifying that the bomb is in certain states is not necessary? Or should a check like that just be its own test entirely?

File: overlord/features/bomb_arming.feature (hunk 5 above)
Comment (jklemen) [inline overlord/features/bomb_arming.feature]: (:

File: overlord/features/bomb_code_submit.feature (hunk 7 above)
Comment (jklemen) [inline overlord/features/bomb_code_submit.feature]: So delete this test?

File: overlord/features/bomb_arming.feature (hunk 4 above)
Comment (djkotowski) [inline overlord/features/bomb_arming.feature]: Those checks should be their own tests (e.g. When I boot the bomb, it should be disarmed)

File: overlord/features/step_definitions/bomb_dis_arming.rb [hunk 15]
```diff
@@ -0,0 +1,52 @@
+# encoding: utf-8
+require './classes/bomb.rb'
```
Comment (djkotowski) [inline overlord/features/step_definitions/bomb_dis_arming.rb]: You shouldn't need this require.

File: overlord/views/index.haml [hunk 16]
```diff
@@ -0,0 +1,51 @@
+!!!
```
Comment (djkotowski) [inline overlord/views/index.haml]: Yay!  Haml!

File: overlord/features/bomb_arming.feature [hunk 17]
```diff
@@ -0,0 +1,25 @@
...
+Feature: Arming the bomb
+  In order to arm the bomb
+  As an evil overlord
+  I should have to enter the correct code
+
+  @happy
+  Scenario: Arming the bomb with a correct code
+    Given I'm at the bomb page
+    And I have already submitted my codes
+    And the bomb is disarmed
```
Comment (djkotowski) [inline overlord/features/bomb_arming.feature]: This is testing the state of the previous two Givens.  If you want to test that the bomb is disarmed on boot, make that a separate test.

File: overlord/features/bomb_arming.feature (hunk 4 above)
Comment (djkotowski) [inline overlord/features/bomb_arming.feature]: I like having the two steps here, as an armed bomb and a timer counting down are potentially two different results.
